cli/command/container/run.go

```diff
@@ -182,37 +182,42 @@ func runContainer(dockerCli command.Cli, opts *runOptions, copts *containerOptio
 		}()
 	}
 	attach := config.AttachStdin || config.AttachStdout || config.AttachStderr
+	statusChan := waitExitOrRemoved(ctx, dockerCli, createResponse.ID, copts.autoRemove)
+	startc := func() error {
+		if err := client.ContainerStart(ctx, createResponse.ID, types.ContainerStartOptions{}); err != nil {
+			// If we have hijackedIOStreamer, we should notify
+			// hijackedIOStreamer we are going to exit and wait
+			// to avoid the terminal are not restored.
+			if attach {
+				cancelFun()
+				<-errCh
+			}
+
+			reportError(stderr, "run", err.Error(), false)
+			if copts.autoRemove {
+				// wait container to be removed
+				<-statusChan
+			}
+			return runStartContainerErr(err)
+		}
+		return nil
+	}
 	if attach {
 		if opts.detachKeys != "" {
 			dockerCli.ConfigFile().DetachKeys = opts.detachKeys
 		}
 
-		close, err := attachContainer(ctx, dockerCli, &errCh, config, createResponse.ID)
+		close, err := attachContainer(ctx, dockerCli, &errCh, config, createResponse.ID, startc)
 
 		if err != nil {
 			return err
 		}
 		defer close()
 	}
 
-	statusChan := waitExitOrRemoved(ctx, dockerCli, createResponse.ID, copts.autoRemove)
-
 	//start the container
-	if err := client.ContainerStart(ctx, createResponse.ID, types.ContainerStartOptions{}); err != nil {
-		// If we have hijackedIOStreamer, we should notify
-		// hijackedIOStreamer we are going to exit and wait
-		// to avoid the terminal are not restored.
-		if attach {
-			cancelFun()
-			<-errCh
-		}
-
-		reportError(stderr, "run", err.Error(), false)
-		if copts.autoRemove {
-			// wait container to be removed
-			<-statusChan
-		}
-		return runStartContainerErr(err)
+	if !attach {
```

Comment:
I guess this `if !attach` can be moved as an `else` to the previous condition?
```go
if attach {
...
} else {
// start the container
...
}

```

Reply:
Thanks, you are right.